src/Linters/PreferSingleQuotedStringLiteralLinter.hack

```diff
@@ -0,0 +1,109 @@
+/*
+ *  Copyright (c) 2017-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the MIT license found in the
+ *  LICENSE file in the root directory of this source tree.
+ *
+ */
+
+namespace Facebook\HHAST;
+
+use namespace HH\Lib\{C, Regex, Str};
+
+final class PreferSingleQuotedStringLiteralLinter extends AutoFixingASTLinter {
+  const type TContext = Script;
+  const type TNode = DoubleQuotedStringLiteralToken;
+
+  // hackfmt-ignore
+  const keyset<string> SINGLE_LETTER_ESCS = keyset[
+    '\\n', '\\r', '\\t', '\\v', '\\e', '\\f'
+  ];
+
+  <<__Override>>
+  public function getLintErrorForNode(
+    this::TContext $_context,
+    this::TNode $node,
+  ): ?ASTLintError {
+    if ($this->isREString($node)) {
+      return null;
+    }
+    $string_contents = $node->getText() |> Str\slice($$, 1, Str\length($$) - 2);
+    if ($this->couldHaveBeenASingleQuotedString($string_contents)) {
+      return new ASTLintError(
+        $this,
+        'Prefer a single quoted string when possible',
+        $node,
+        () ==> $this->toSingleQuotedString($node),
+      );
+    }
+
+    return null;
+  }
+
+  private function isREString(this::TNode $node): bool {
+    $ast = $this->getAST();
+    return $ast->getParentOfDescendant($node)
+      |> $ast->getParentOfDescendant($$) is PrefixedStringExpression;
+  }
+
+  /**
+   * @see https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.double
+   *
+   * @license The inline definitions of the Regex patterns come from the PHP manual.
+   *          They fall under the license of the php.net website.
+   *          https://www.php.net/license/index.php
+   *          As of the time of writing, this is the https://creativecommons.org/licenses/by/3.0/legalcode
```

Comment:
We're unable to take this, and it can't be contributed under Facebook's CLA unless you own the copyright.

Reply:
I wasn't quite sure about this, that is why I put it in code.

This applies to the inline comments that refer back to the spec of the double quoted string literal.
The regex patterns themselves must be functionally equivalent.

Would removing the inline comments (and the removal of the `@license`) make this acceptable under the CLA?

Reply:
> The regex patterns themselves must be functionally equivalent.

Is there a more liberal form that can be done? We don't need to identify exactly what substring is an escape sequence, merely that the string contains an escape sequence.

Reply:
I could say `re"!\\\\{?[0-9a-f]!i"`.
This would capture any backslash followed by a valid hex character with one optional `{` in between.
This would have more false negatives, but it is not a license problem.
_(Also more perf, since there is only one regex)._

Reply:
Sounds good to me :)

Reply:
Fixed